### PR TITLE
MLOPS-306 - MS2DeepScore Refactor

### PR DESCRIPTION
**Pull request by pedrocwb**
This PR merges both  [MLOPS-306: MS2DeepScore notebook](https://github.com/omigami/romigami/pull/6) and [MLOPS-306: Add MS2DeepScore](https://github.com/omigami/romigami/pull/4).

Instead of wrapping the API in single functions, we let the user access the API choosing which client he/she wants to get predictions of. 

```
omigami <- omigami_init()
client <- omigami$Spec2Vec(token="my_token") 

client$match_spectra_from_path(**args)

```

### CODE REVIEW

File: notebook/ms2deepscore/tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    "*In order to use omigami, you need an access token provided by DataRevenue.*\n",
+    "## Introduction\n",
+    "\n",
+    "This is a short guide on how you can predict molecular structural similarities (Tanimoto scores) from pairs of mass spectrometry spectra in local mgf files - using Omigami's MS2DeepScore API.\n",
+    "\n",
+    "The API uses a MS2DeepScore model which was trained on > 100,000 MS/MS spectra from GNPS. It prepares the two spectra and using a Siamese neural network predicts returns the molecular structural similarity (Tanimoto score).\n",
+    "\n",
+    "This notebook shows you how to:\n",
+    "1. Specify any MGF file with two spectra.\n",
+    "2. Make a prediction through the MS2DeepScore API."
```
Comment (dianacarvalho1) [inline notebook/ms2deepscore/tutorial.ipynb]: This is not updated. Please copy from the python notebook

File: notebook/ms2deepscore/tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+   "metadata": {},
+   "source": [
+    "# 2. Predict molecular structural similarity with MS2DeepScore\n",
+    "\n",
+    "`Romigami` is a R wrapper which instantiates a virtual environment on which to install and call the Python `Omigami` package\n",
+    "\n",
+    "`MS2DeepScore` is a python wrapper which :\n",
+    "- Builds a json payload from the MGF file\n",
+    "- Calls the MS2DeepScores API\n",
+    "- Formats the prediction results into readable format"
```
Comment (dianacarvalho1) [inline notebook/ms2deepscore/tutorial.ipynb]: we could include a saving part like in spec2vec too

File: notebook/spec2vec/tutorial.ipynb [hunk 3]
```diff
@@ -82,8 +82,10 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
-   "metadata": {},
+   "execution_count": 28,
+   "metadata": {
+    "tags": []
```
Comment (dianacarvalho1) [inline notebook/spec2vec/tutorial.ipynb:92]: in the spec2vec notebook it misses to talk about the `metadata_list` and the `ion_mode`. honestly I think the "talk" we can literally copy from the python notebooks

File: notebook/spec2vec/tutorial.ipynb [hunk 4]
```diff
@@ -284,8 +180,7 @@
     "    dir.create(\"matches\")\n",
     "}\n",
     "for (i in seq_along(spectra_matches)){\n",
-    "    matches <- data.frame(lapply(spectra_matches[i], as.character), stringsAsFactors=FALSE)\n",
-    "    write.csv(matches, sprintf(\"matches/spectrum_%s.csv\", i))\n",
+    "    write.csv(spectra_matches[[i]],sprintf(\"matches/spectrum_%s.csv\",i))\n",
```
Comment (dianacarvalho1) [inline notebook/spec2vec/tutorial.ipynb]: are you sure this works?? last time I tried I had to make those weird adjustments that are on the diff

File: vignettes/ms2deepscore/tutorial.Rmd [hunk 5]
```diff
@@ -0,0 +1,75 @@
+---
+title: An R Markdown document converted from "notebook/ms2deepscore/tutorial.ipynb"
+output: html_document
+---
+
```
Comment (dianacarvalho1) [inline vignettes/ms2deepscore/tutorial.Rmd]: once the notebook is updated, this should be updated too

File: vignettes/spec2vec/tutorial.Rmd [hunk 6]
```diff
@@ -68,16 +68,18 @@ library("romigami")
 ```
 
 ```{r}
-# environment setup
-initialize_environment()
+# Initialize omigami environment
+omigami <- omigami_init()
 ```
```
Comment (dianacarvalho1) [inline vignettes/spec2vec/tutorial.Rmd]: this one as well. particularly the explanations part

**Review by dianacarvalho1 — COMMENTED** [review]
Thanks @pedrocwb ! I think it looks better but there are still somethings to polish. and we also shouldn't merge this until we get the endpoint up and running. Feel free to close the other PRs (I think you joined everything here right?)

File: notebook/spec2vec/tutorial.ipynb (hunk 4 above)
Comment (pedrocwb) [reply notebook/spec2vec/tutorial.ipynb]: Nah, Idk why this changed.

**Comment by pedrocwb** [comment]
> close the other PRs (I think you joined everything here right?)

yes. I joined everything here. All comments were addressed. 

Thank you @dianacarvalho1

File: README.md [hunk 7]
```diff
@@ -69,9 +73,38 @@ The supported metadata keys for omigami are (case insensitive):
...
+                                   include_metadata = include_metadata,
+                                   ion_mode = ion_mode
+)
+  ```
+
+#### Notebooks
+You can find a [tutorial](https://github.com/omigami/romigami/blob/master/notebook/ms2deepscore/tutorial.ipynb) notebook in the `/notebooks/` folder.
+
+And an [R Markdown document](https://github.com/omigami/romigami/blob/master/ms2deepscore/spec2vec/tutorial.Rmd) on the `/vignettes/` folder.
+
```
Comment (PierreMkt) [inline README.md]: The Notebooks section is duplicated

File: notebook/spec2vec/tutorial.ipynb [hunk 8]
```diff
@@ -63,73 +63,107 @@
...
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 4,
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "client <- omigami$Spec2Vec(token=\"73gyPBFLyA5zp1WYh0XzGBLMCiilJ3Gr\")"
```
Comment (PierreMkt) [inline notebook/spec2vec/tutorial.ipynb]: the token should be removed from the jupyter notebook and the R markdown files right ?

**Review by PierreMkt — APPROVED** [review]
looks good to me. Just a couple comments :)

File: notebook/spec2vec/tutorial.ipynb (hunk 8 above)
Comment (pedrocwb) [reply notebook/spec2vec/tutorial.ipynb]: good catch.

File: README.md (hunk 7 above)
Comment (pedrocwb) [reply README.md]: Actualy one is for the MS2Deep and another one for Spec2vec. 

They are in different sessions with different links

File: README.md [hunk 9]
```diff
@@ -69,9 +73,38 @@ The supported metadata keys for omigami are (case insensitive):
 - "inchikey_inchi"
 
 #### Notebooks
-You can find a [tutorial](https://github.com/omigami/romigami/blob/master/notebook/Tutorial.ipynb) notebook in the `/notebooks/` folder.
+You can find a [tutorial](https://github.com/omigami/romigami/blob/master/notebook/spec2vec/tutorial.ipynb) notebook in the `/notebooks/` folder.
+
+And an [R Markdown document](https://github.com/omigami/romigami/blob/master/vignettes/spec2vec/tutorial.Rmd) on the `/vignettes/` folder.
```
Comment (dianacarvalho1) [inline README.md:78]: why do we need this big path with blob/master/ ?

File: README.md [hunk 10]
```diff
@@ -69,9 +73,38 @@ The supported metadata keys for omigami are (case insensitive):
...
+                                   n_best = n_best_matches,
+                                   include_metadata = include_metadata,
+                                   ion_mode = ion_mode
+)
+  ```
+
+#### MS2DeepScore Notebooks
+You can find a [tutorial](https://github.com/omigami/romigami/blob/master/notebook/ms2deepscore/tutorial.ipynb) notebook in the `/notebooks/` folder.
+
+And an [R Markdown document](https://github.com/omigami/romigami/blob/master/vignettes/ms2deepscore/tutorial.Rmd)) on the `/vignettes/` folder.
```
Comment (dianacarvalho1) [inline README.md:105]: here too

File: README.md [hunk 11]
```diff
@@ -82,9 +115,16 @@ And an [R Markdown document](https://github.com/omigami/romigami/blob/master/vig
 5. These embeddings will be compared against the reference embeddings around the Precursor MZ.
 6. The N best matches per spectrum are returned on the response as pandas dataframes.  
 
+### MS2DeepScore
+1. Save your pair of spectra data in a MGF file locally
```
Comment (dianacarvalho1) [inline README.md]: why pair? it can be as many as the user wants

File: README.md [hunk 12]
```diff
@@ -82,9 +115,16 @@ And an [R Markdown document](https://github.com/omigami/romigami/blob/master/vig
...
+2. Create an MS2DeepScore with your user token
+3. Call `match_spectra_from_path` with the location of your mgf file.
+4. The MGF spectra data will be processed and sent to the trained neural network that will predict the molecular structural similarity. 
+5. The prediction is returned on the response as a dictionary.
```
Comment (dianacarvalho1) [inline README.md]:
```suggestion
4. The MGF spectra data will be processed and sent to the trained neural network that will predict the molecular structural similarity against reference spectra around the Precursor MZ.
5. The N best matches per spectrum are returned on the response as pandas dataframes.  
```

File: notebook/ms2deepscore/tutorial.ipynb [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    "\n",
+    "\n",
+    "`MS2DeepScore` is a python wrapper which :\n",
+    "- Builds a json payload from the MGF file\n",
+    "- Calls the MS2DeepScores API\n",
+    "- Formats the prediction results into readable dataframes\n",
+    "\n",
+    "____\n",
+    "- `n_best` sets the number of matches you'd like per spectrum (it is set to 10 by default).\n",
+    "- `include_metadata` selects the metadata fields that will be displayed in the resulting top matches.\n",
```
Comment (dianacarvalho1) [inline notebook/ms2deepscore/tutorial.ipynb:62]: misses the ion mode

File: tests/assets/GNPS-2-EQUAL-SPECTRA.mgf [hunk 14]
```diff
@@ -0,0 +1,73 @@
+BEGIN IONS
+PEPMASS=377.232
+CHARGE=1
```
Comment (dianacarvalho1) [inline tests/assets/GNPS-2-EQUAL-SPECTRA.mgf]: is this file still necessary?

File: tests/testthat/test_ms2deepscore.R [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+            )
+  expect_equal(results["Tanimoto Score"], 1)
+})
+
+test_that("token is invalid", {
+
+  mgf_path <- paste0(dirname(getwd()), "/assets/GNPS-COLLECTIONS-MISC.mgf")
+
+  omigami <- omigami_init()
+  client <- omigami$MS2DeepScore(token="Randon Token")
```
Comment (dianacarvalho1) [inline tests/testthat/test_ms2deepscore.R]: *Random

File: tests/testthat/test_spec2vec.R [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+  expect_equal(length(results), 46)
+})
+
+test_that("token is invalid", {
+
+  mgf_path <- paste0(dirname(getwd()), "/assets/GNPS-COLLECTIONS-MISC.mgf")
+
+  omigami <- omigami_init()
+  client <- omigami$Spec2Vec(token="Randon Token")
```
Comment (dianacarvalho1) [inline tests/testthat/test_spec2vec.R]: *Random

File: vignettes/ms2deepscore/tutorial.Rmd [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+`MS2DeepScore` is a python wrapper which :
+- Builds a json payload from the MGF file
+- Calls the MS2DeepScores API
+- Formats the prediction results into readable dataframes
+
+____
+- `n_best` sets the number of matches you'd like per spectrum (it is set to 10 by default).
+- `include_metadata` selects the metadata fields that will be displayed in the resulting top matches.
+
```
Comment (dianacarvalho1) [inline vignettes/ms2deepscore/tutorial.Rmd:45]: ion_mode

**Review by dianacarvalho1 — COMMENTED** [review]
Got a few more comments @pedrocwb ! I know I'm being a pain, but I think these public repos should be squeaky clean. And also, I feel that the format of the code is weird in places.. can we do something like Black in R? Just to make everything pretty

File: README.md (hunk 9 above)
Comment (pedrocwb) [reply README.md:78]: We need it to link to the file in the repo.

**Review by dianacarvalho1 — APPROVED** [review]
